File: rails/app/models/achievement.rb
class Achievement < ApplicationRecord
  self.table_name = 'achievements'

  enum :rarity, { common: 0, uncommon: 1, rare: 2, epic: 3, legendary: 4 }

  validates :name, presence: true, length: { maximum: 200 }

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:points_positive, 'Achievement must award at least one point') unless ((points.nil? || points > 0))
  end

  def to_s
    name.to_s
  end

  # Business operations

  def point_value(multiplier)
    # TODO: implement point_value
    nil
  end

  def reveal
    # TODO: implement reveal
  end
end

File: dotnet/CardsProject/Domain/Players/Achievement.cs
using System.ComponentModel.DataAnnotations;

namespace CardsProject.Domain.Players;

public enum AchievementRarityType
{
    Common,
    Uncommon,
    Rare,
    Epic,
    Legendary
}

public class Achievement : IValidatableObject
{
    public int Id { get; set; }

    public string Name { get; set; } = "";
    public string Description { get; set; } = "";
    public string? IconUrl { get; set; }
    public int Points { get; set; } = 10;
    public AchievementRarityType Rarity { get; set; }
    public bool IsHidden { get; set; } = false;

    // Business operations

    public int PointValue(int multiplier)
    {
        // TODO: implement point_value
        return default;
    }

    public void Reveal()
    {
        // TODO: implement reveal
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( Points > 0 ))
            yield return new ValidationResult("Achievement must award at least one point", new[] { nameof(Id) });
    }
}
